ERROR: spdlog  error C2338: Cannot format an argument. To make type T formattable provide a formatter<T> specialization:

Thanks for your bug report. We appreciate it a lot.

win10 + VS2019
I use cmake-gui to generate the solution and then started to compile the file, and got the following problem,
1>------ 已启动生成: 项目: mmdeploy_core, 配置: Debug x64 ------
1>tensor.cpp
1>E:\vOpenMM\mmdeploy\csrc\core/status_code.h(98): warning C4275: 非 dll 接口 class“system_error2::status_code_domain”用作 dll 接口 class“mmdeploy::StatusDomain”的基
1>E:\vOpenMM\mmdeploy\third_party\outcome\outcome-experimental.hpp(6598): note: 参见“system_error2::status_code_domain”的声明
1>E:\vOpenMM\mmdeploy\csrc\core/status_code.h(98): note: 参见“mmdeploy::StatusDomain”的声明
1>E:\vOpenMM\spdlog\include\spdlog/fmt/bundled/core.h(1728): error C2338: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
1>E:\vOpenMM\spdlog\include\spdlog/fmt/bundled/core.h(1849): note: 查看对正在编译的函数 模板 实例化“fmt::v8::detail::value<Context> fmt::v8::detail::make_arg<true,Context,fmt::v8::detail::type::custom_type,mmdeploy::DataType&,0>(T)”的引用
1>        with
1>        [
1>            Context=fmt::v8::format_context,
1>            T=mmdeploy::DataType &
1>        ]
1>E:\vOpenMM\spdlog\include\spdlog/fmt/bundled/core.h(1871): note: 查看对正在编译的函数 模板 实例化“fmt::v8::format_arg_store<fmt::v8::format_context,mmdeploy::DataType,mmdeploy::DataType>::format_arg_store<mmdeploy::DataType&,const mmdeploy::DataType&>(mmdeploy::DataType &,const mmdeploy::DataType &)”的引用
1>E:\vOpenMM\spdlog\include\spdlog/fmt/bundled/core.h(1872): note: 查看对正在编译的函数 模板 实例化“fmt::v8::format_arg_store<fmt::v8::format_context,mmdeploy::DataType,mmdeploy::DataType>::format_arg_store<mmdeploy::DataType&,const mmdeploy::DataType&>(mmdeploy::DataType &,const mmdeploy::DataType &)”的引用
1>E:\vOpenMM\spdlog\include\spdlog/logger.h(357): note: 查看对正在编译的函数 模板 实例化“fmt::v8::format_arg_store<fmt::v8::format_context,mmdeploy::DataType,mmdeploy::DataType> fmt::v8::make_format_args<fmt::v8::format_context,mmdeploy::DataType&,const mmdeploy::DataType&>(mmdeploy::DataType &,const mmdeploy::DataType &)”的引用
1>E:\vOpenMM\spdlog\include\spdlog/logger.h(89): note: 查看对正在编译的函数 模板 实例化“void spdlog::logger::log_<mmdeploy::DataType&,const mmdeploy::DataType&>(spdlog::source_loc,spdlog::level::level_enum,spdlog::string_view_t,mmdeploy::DataType &,const mmdeploy::DataType &)”的引用
1>E:\vOpenMM\mmdeploy\csrc\core\tensor.cpp(99): note: 查看对正在编译的函数 模板 实例化“void spdlog::logger::log<mmdeploy::DataType&,const mmdeploy::DataType&>(spdlog::source_loc,spdlog::level::level_enum,fmt::v8::basic_format_string<char,mmdeploy::DataType &,const mmdeploy::DataType &>,mmdeploy::DataType &,const mmdeploy::DataType &)”的引用
1>registry.cpp
1>E:\vOpenMM\mmdeploy\csrc\core/status_code.h(98): warning C4275: 非 dll 接口 class“system_error2::status_code_domain”用作 dll 接口 class“mmdeploy::StatusDomain”的基
1>E:\vOpenMM\mmdeploy\third_party\outcome\outcome-experimental.hpp(6598): note: 参见“system_error2::status_code_domain”的声明
1>E:\vOpenMM\mmdeploy\csrc\core/status_code.h(98): note: 参见“mmdeploy::StatusDomain”的声明
1>stacktrace.cpp
1>正在生成代码...
1>已完成生成项目“mmdeploy_core.vcxproj”的操作 - 失败。

currently I'm not sure whether this is a spdlog problem or a mmdeploy problem or both. 
the only thing I'm sure is that this error is raised by the following sentence, in tensor.cpp.
```
MMDEPLOY_ERROR("mismatched data type {} vs {}", desc_.data_type, tensor.desc().data_type);
```
My temporary fix is make them integer,
```
MMDEPLOY_ERROR("mismatched data type {} vs {}", (int)desc_.data_type, (int)tensor.desc().data_type);
```
any better idea, please?

--- a/csrc/core/utils/formatter.h
+++ b/csrc/core/utils/formatter.h
@@ -3,10 +3,14 @@
 #ifndef MMDEPLOY_SRC_UTILS_FORMATTER_H_
 #define MMDEPLOY_SRC_UTILS_FORMATTER_H_
 
+#include <ostream>
+
 #include "core/logger.h"
+#include "core/types.h"
 
 #if FMT_VERSION >= 50000
 #include "spdlog/fmt/bundled/ranges.h"
+#include "spdlog/fmt/ostr.h"
 #else
 #include <type_traits>
 #endif
@@ -17,12 +21,55 @@ class Value;
 
 MMDEPLOY_API std::string format_value(const Value& value);
 
+inline std::string to_string(PixelFormat format) {
+  switch (format) {
+    case PixelFormat::kBGR:
+      return "BGR";
+    case PixelFormat::kRGB:
+      return "RGB";
+    case PixelFormat::kGRAYSCALE:
+      return "GRAYSCALE";
+    case PixelFormat::kNV12:
+      return "NV12";
+    case PixelFormat::kNV21:
+      return "NV21";
+    case PixelFormat::kBGRA:
+      return "BGRA";
+    default:
+      return "invalid_format_enum";
+  }
+}
+
+inline std::string to_string(DataType type) {
+  switch (type) {
+    case DataType::kFLOAT:
+      return "FLOAT";
+    case DataType::kHALF:
+      return "HALF";
+    case DataType::kINT8:
+      return "INT8";
+    case DataType::kINT32:
+      return "INT32";
+    case DataType::kINT64:
+      return "INT64";
+    default:
+      return "invalid_data_type_enum";
+  }
+}
+
+inline std::ostream& operator<<(std::ostream& os, PixelFormat format) {
+  return os << to_string(format);
+}
+
+inline std::ostream& operator<<(std::ostream& os, DataType type) { return os << to_string(type); }
+
 }  // namespace mmdeploy
 
 namespace fmt {
 
 #if FMT_VERSION >= 50000
 
+// `Value` maybe an incomplete type at this point, making `operator<<` not usable
 template <>
 struct formatter<mmdeploy::Value> {
   constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }

--- a/csrc/core/utils/formatter.h
+++ b/csrc/core/utils/formatter.h
@@ -81,16 +81,16 @@ struct formatter<mmdeploy::Value> {
 
 #else
 
-inline void format_arg(BasicFormatter<char> &f, const char *, const mmdeploy::Value &d) {
+inline void format_arg(BasicFormatter<char>& f, const char*, const mmdeploy::Value& d) {
   f.writer() << mmdeploy::format_value(d);
 }
 
 template <typename T>
-auto format_arg(BasicFormatter<char> &f, const char *, const T &v)
+auto format_arg(BasicFormatter<char>& f, const char*, const T& v)
     -> std::void_t<decltype(begin(v), end(v))> {
   f.writer() << "[";
   bool first = true;
-  for (const auto &x : v) {
+  for (const auto& x : v) {
     f.writer() << (first ? "" : ", ") << fmt::format("{}", x);
     first = false;
   }

--- a/csrc/core/utils/formatter.h
+++ b/csrc/core/utils/formatter.h
@@ -7,10 +7,10 @@
 
 #include "core/logger.h"
 #include "core/types.h"
+#include "spdlog/fmt/ostr.h"
 
 #if FMT_VERSION >= 50000
 #include "spdlog/fmt/bundled/ranges.h"
-#include "spdlog/fmt/ostr.h"
 #else
 #include <type_traits>
 #endif
@@ -83,11 +83,6 @@ struct formatter<mmdeploy::Value> {
 
 inline void format_arg(BasicFormatter<char> &f, const char *, const mmdeploy::Value &d) {
   f.writer() << mmdeploy::format_value(d);
-}
-
-template <typename T, std::enable_if_t<std::is_enum<std::decay_t<T> >::value, bool> = true>
-void format_arg(BasicFormatter<char> &f, const char *, const T &v) {
-  f.writer() << (int)v;
 }
 
 template <typename T>

